### PR TITLE
Keep download archive in memory for better performance

### LINKED ISSUE #122: Sort the download archive for better performance
My YTDL archive is quite large:

```
$ wc -l ytdl_archive.txt
76923 ytdl_archive.txt
```

I have noticed that performance for "is the video in the archive?" checking becomes slow quickly, even if only 100 more video lines were added to an already-sorted archive, but a quick command-line sort of the archive results in a massive performance boost:

```
$ sort < ytdl_archive.txt > xxx && mv xxx ytdl_archive.txt
```

I would suggest that for any batch (channel, playlist, etc.) download that also uses the archive and adds at least one line to the archive, sort the archive at the end of the batch to maintain performance for future instances that use that archive. It may also be good to add an option to disable this post-batch archive sorting for anyone who also uses the archive as a record of the original order of downloads.

### PR DESCRIPTION
**Pull request by jbruchon**
- I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

Fixes #122

### CODE REVIEW

**Comment by jbruchon** [comment]
OK, hold on...I'm having a weird problem. I'm seeing it work for about 4 videos in this particular set, then it starts trying to mux already-muxed files and whatnot. It shouldn't be doing that, but I don't understand why it would work for multiples and then just stop working like that.

**Comment by jbruchon** [comment]
Working properly for me now. Should be ready for merge.

**Comment by blackjack4494** [comment]
Will review it later. The CI is failing because of my hotstar fix where I forgot one case for the regex so it's failing. 
As for your test with 70k lines how is the ram usage?  

As an alternative some basic in-file indexing could be implemented. Then you won't store all data in memory (or you basically can select specific regions). Simple approach is sorting for start letter and group those. Then in the very first line start with some identifier like `#` and simply list the index of the letter.  
Just dropping this here before I forget this 😅

**Comment by jbruchon** [comment]
I don't know how to officially test the peak RAM usage for Python, but since we're talking about short (~20 byte including newline) strings, it's probably not so bad. I have it running now (it's in the middle of a download) and it says 114MB, and that's with the 77251-item archive loaded. The archive itself is 1.6 MB in size. If this were C, each entry in the BST would consume about 36 bytes. The first list should be garbage collected upon completion of the preload function.

**Review by blackjack4494 — APPROVED** [review]
looks good.

**Comment by blackjack4494** [comment]
Will merge it when Travis is done. Seems to have some issues lately with multiple builds spawning at once. A new release should be coming later today. Want to include some more features for next one.

File: youtube_dlc/YoutubeDL.py [hunk 1]
```diff
@@ -113,6 +113,53 @@
 if compat_os_name == 'nt':
     import ctypes
 
+# Archive tree
+class ArchiveTree(object):
+    def __init__(self, line):
+        self.left = None
+        self.right = None
+        self.line = line
```
Comment (blackjack4494) [inline youtube_dlc/YoutubeDL.py:121]: You could remove the comment and add this under the class `"""Archive tree"""` or some very short description

**Review by blackjack4494 — CHANGES_REQUESTED** [review]

**Comment by blackjack4494** [comment]
@jbruchon you missed one blank line at `117`. Try running `flake8 .`  
It expects 2 blank before a new `class`. If you add a comment it won't count as a blank line anymore.

File: youtube_dlc/YoutubeDL.py (hunk 1 above)
Comment (jbruchon) [reply youtube_dlc/YoutubeDL.py]: Done and done.

**Comment by jbruchon** [comment]
I fixed it. I ran it and got no messages:

```
C:\github\youtube-dlc>flake8
.\youtube_dlc\YoutubeDL.py:117:1: E302 expected 2 blank lines, found 1

C:\github\youtube-dlc>flake8

C:\github\youtube-dlc>
```

**Review by blackjack4494 — APPROVED** [review]

**Comment by blackjack4494** [comment]
@jbruchon is it worth to add some kind of rate limiting for checking archive?

**Comment by jbruchon** [comment]
You mean intentionally slow down the program for no reason?

**Comment by blackjack4494** [comment]
Some sites like youtube will return an error if you send too many requests within a certain time (your missing reason). So I thought a way to limit this would be a good feature.

**Comment by jbruchon** [comment]
The archive check doesn't send any requests though.

**Pull request merged**